### PR TITLE
enhancement!: validate input in content API create and update controllers

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- `coreController` `create` and `update` methods call `validateInput()`
- `validateInput` now throws on fields that don't exist on schema
- fixes bug in `validateInput` where only the first traversal method was being called

### Why is it needed?

To make it easier to debug content API `create` and `update` requests; we should not be silently removing anything from the users input data.

### How to test it?

API tests have been added

To test manually, create a content type and try sending various types of bad input to it (create and update endpoints), and you should receive a 400 bad request instead of a 200/201 success.

### Related issue(s)/PR(s)

DX-1136

### CODE REVIEW

**Comment by innerdvations** [comment]
@Boegie19 I know you're interested in this, so any thoughts or comments? I've added additional details to the error response, but if there's anything else you would find helpful that can be added easily, let me know!

**Comment by Boegie19** [comment]
I love the improved error messages. That realy helps.
I have 2 questions.
1 Lets say I want to give an API parameter to active a document middleware or an before Event to the query engine. Will this still work. 
2 the new error types should be documented for plugin developers since plugins need to wrap errors happening inside there own strapi side and client side components 

> anything else you think would be helpful for api consumers that can be added easily

On the top of my head there is not else that can easily be added that would.

Only other thing would be someting like the OPEN API plugin where deep filter, fields and populate are documented correctly. What would help with direct API access or an SDK with auto complete. for that.

**Comment by innerdvations** [comment]
I'll make sure this response format gets documented.

Could you give an example of 1?

**Comment by innerdvations** [comment]
Adding `don't merge` label for now until I can be sure we won't be breaking a lot more things than expected by preventing unrecognized attributes.

**Comment by Marc-Roig** [comment]
Any middleware can extend the context in itself, in this case I would extend the context.params which is the one that includes, fields, filters, ...

File: packages/core/core/src/core-api/controller/index.ts [hunk 1]
```diff
@@ -59,6 +59,7 @@ function createController({
       return strapi.contentAPI.validate.query(ctx.query, contentType, { auth });
     },
 
+    // TODO: should this just accept ctx and pull body.data from it?
```
Comment (Convly) [inline packages/core/core/src/core-api/controller/index.ts:62]: We might want to validate other data than ctx.body, I think it makes sense to have both parameters

File: packages/core/utils/src/validate/visitors/throw-unrecognized-fields.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  if (
+    isComponentSchema(schema) &&
+    isDynamicZoneAttribute(parent?.attribute) &&
+    DYNAMIC_ZONE_ALLOWED_FIELDS.includes(key)
+  ) {
+    return;
+  }
+
+  // allow special relation reordering keys
+  // TODO: Only toMany or all?
```
Comment (Convly) [inline packages/core/utils/src/validate/visitors/throw-unrecognized-fields.ts:49]: @Marc-Roig is relation reordering enabled for toMany relations only? Or should it be possible to have it for any kind relation?

Comment (Marc-Roig) [reply packages/core/utils/src/validate/visitors/throw-unrecognized-fields.ts:49]: Only manytoX and XToMany relations have ordering

**Review by Bassel17 — APPROVED** [review]
QAed!
LGTM

**Pull request merged**